File: face-modifications/face_modifications.py
from PIL import Image
import imagehash
import numpy as np
import pandas as pd
import math, os

SIZE = 8

def process_image_pair(fakeimg, realimg, p):
    width, height = fakeimg.size

    # Calculate the width and height of each subrectangle
    subwidth = math.ceil(width / p)
    subheight = math.ceil(height / p)

    ans, cnt = 0, 0

    for i in range(p):
        for j in range(p):
            # Calculate subrectangle boundaries
            left = j * subwidth
            upper = i * subheight
            right = min((j + 1) * subwidth, width)  # Ensure we don't exceed image width
            lower = min((i + 1) * subheight, height)  # Ensure we don't exceed image height

            # Crop subrectangles
            f_sub = fakeimg.crop((left, upper, right, lower))
            r_sub = realimg.crop((left, upper, right, lower))

            # Compute perceptual hashes using the SIZE constant
            f_hash = imagehash.phash(f_sub, hash_size=SIZE)
            r_hash = imagehash.phash(r_sub, hash_size=SIZE)

            # Compute hash difference
            dist = f_hash - r_hash
            if dist > ans:
                ans = dist
            cnt +=1

    return ans, cnt

l = [1,2,3,4,5,6]
result = {}

tot_path = "C:/Users/HP/Downloads/images/"

path_names = [ "m1", "m2", "f1", "f2" ]


# iterate over pathnames
for name in path_names:
    result = {}
    # now there are a bunch of folders
    # iterate over all except base
    for iter in ["eyes", "nose", "nose_code", "lips", "lips_code"]:
        li = []
        if iter.find("eyes")==-1:
            st = ""
            if iter.find("lips")!=-1:
                st="lip"
            else:
                st="nose"
            # we have to compare base and nosebox
            real_path = tot_path + name +'/'+ 'base' +'/base.jpg'
            fake_path = tot_path + name +'/'+ iter +'/base.jpg'
            real = Image.open(real_path)
            fake = Image.open(fake_path)
            d, cnt = process_image_pair(fake, real, 1)
            li.append(d)

            for p in range(1,6):
                real_path = tot_path + name +'/'+ 'base' +f'/{st}box.jpg'
                fake_path = tot_path + name +'/'+ iter +f'/{st}box.jpg'
                real = Image.open(real_path)
                fake = Image.open(fake_path)
                d, cnt = process_image_pair(fake, real, p)
                li.append(d)
        else:
            # we have to compare base and eyeboxes
            real_path = tot_path + name +'/'+ 'base' +'/base.jpg'
            fake_path = tot_path + name +'/'+ iter +'/base.jpg'
            real = Image.open(real_path)
            fake = Image.open(fake_path)
            d, cnt = process_image_pair(fake, real, 1)
            li.append(d)

            for p in range(1,6):
                x = 0
                for num in range(2):
                    real_path = tot_path + name +'/'+ 'base' +f'/eyebox_{num}.jpg'
                    fake_path = tot_path + name +'/'+ iter +f'/eyebox_{num}.jpg'
                    real = Image.open(real_path)
                    fake = Image.open(fake_path)
                    d, cnt = process_image_pair(fake, real, p)
                    if x<d:
                        x=d
                li.append(x)

        result[iter] = li
    df = pd.DataFrame(result)
    df.to_csv(f'{name}_face.csv', index=False)

    

File: face-modifications/face_modifications.scala
import java.io.{ByteArrayInputStream, ByteArrayOutputStream, File, FileInputStream, FileOutputStream, PrintWriter}
import java.nio.file.{Files, Paths}
import java.awt.image.BufferedImage
import javax.imageio.ImageIO
import scala.math.{ceil, min}
import scala.collection.mutable.ListBuffer
import java.io.{FileWriter, BufferedWriter, PrintWriter}

// The scala code write incorrectly to csv
// We must carry out a transpose operation on the data
// It writes a column as a row, however the provided csv files have been corrected


object PHashExample extends App {
  val pHash = new ImagePHashScala(size = 64, smallerSize = 8)

  def bufferedImageToInputStream(img: BufferedImage): ByteArrayInputStream = {
    val baos = new ByteArrayOutputStream()
    ImageIO.write(img, "jpg", baos)
    new ByteArrayInputStream(baos.toByteArray)
  }

  def processImagePair(fakeImg: BufferedImage, realImg: BufferedImage, p: Int): Int = {
    val width = fakeImg.getWidth
    val height = fakeImg.getHeight

    val subwidth = ceil(width.toDouble / p).toInt
    val subheight = ceil(height.toDouble / p).toInt

    var maxd = 0

    for (i <- 0 until p; j <- 0 until p) {
      val left = j * subwidth
      val upper = i * subheight
      val right = min((j + 1) * subwidth, width)
      val lower = min((i + 1) * subheight, height)

      val fSub = fakeImg.getSubimage(left, upper, right - left, lower - upper)
      val rSub = realImg.getSubimage(left, upper, right - left, lower - upper)

      val fSubInputStream = bufferedImageToInputStream(fSub)
      val rSubInputStream = bufferedImageToInputStream(rSub)

      val fHash = pHash.getHash(fSubInputStream)
      val rHash = pHash.getHash(rSubInputStream)

      val dist = pHash.hammingDistance(fHash, rHash)
      if(dist>maxd) maxd=dist

      fSubInputStream.close()
      rSubInputStream.close()
    }
    maxd
  }

  val basePath = "C:/Users/HP/Downloads/images/"
  val pathNames = List("m1", "m2", "f1", "f2")
  val iterations = List("eyes", "nose", "nose_code", "lips", "lips_code")
  val outputCsvTemplate = "%s_face.csv"

  def processImagesForName(name: String): Unit = {
    val result = new ListBuffer[String]()

    for (iter <- iterations) {
      val distances = new ListBuffer[Int]()
      val baseImgPath = s"$basePath$name/base/base.jpg"
      val baseImg = ImageIO.read(new File(baseImgPath))

      if (iter.contains("eyes")) {
        for (p <- 1 to 5) {
          var maxDist = 0
          for (num <- 0 until 2) {
            val realImgPath = s"$basePath$name/base/eyebox_$num.jpg"
            val fakeImgPath = s"$basePath$name/$iter/eyebox_$num.jpg"
            val realImg = ImageIO.read(new File(realImgPath))
            val fakeImg = ImageIO.read(new File(fakeImgPath))
            val dist = processImagePair(fakeImg, realImg, p)
            if (dist > maxDist) maxDist = dist
          }
          distances += maxDist
        }
      } else {
        for (p <- 1 to 5) {
          val st = if (iter.contains("lips")) "lip" else "nose"
          val realImgPath = s"$basePath$name/base/${st}box.jpg"
          val fakeImgPath = s"$basePath$name/$iter/${st}box.jpg"
          val realImg = ImageIO.read(new File(realImgPath))
          val fakeImg = ImageIO.read(new File(fakeImgPath))
          val dist = processImagePair(fakeImg, realImg, p)
          distances += dist
        }
      }
      result += distances.mkString(",")
    }

    val outputCsv = outputCsvTemplate.format(name)
    val writer = new PrintWriter(new BufferedWriter(new FileWriter(outputCsv)))
    writer.println("eyes,nose,nose_code,lips,lips_code")
    for (line <- result) {
      writer.println(line)
    }
    writer.close()
  }

  for (name <- pathNames) {
    processImagesForName(name)
  }
}